tools/ci/pr_preview.py

```diff
@@ -0,0 +1,403 @@
+    API, ensuring that the subsequent request will not deplete the relevant
+    allowance. This verification does not itself influence rate limiting:
+
+    > Accessing this endpoint does not count against your REST API rate limit.
+
+    https://developer.github.com/v3/rate_limit/
+    '''
+    def guard_decorator(func):
+        def wrapped(self, *args, **kwargs):
+            limits = gh_request('GET', '{}/rate_limit'.format(self._host))
+
+            values = limits['resources'].get(resource)
+
+            remaining = values['remaining']
+            limit = values['limit']
+
+            logger.info(
+                'Limit for "%s" resource: %s/%s', resource, remaining, limit
+            )
+
+            if limit and float(remaining) / limit < API_RATE_LIMIT_THRESHOLD:
+                raise Exception(
+                    'Exiting to avoid GitHub.com API request throttling.'
+                )
+
+            return func(self, *args, **kwargs)
+        return wrapped
+    return guard_decorator
+
+class Project(object):
+    def __init__(self, host, github_project):
+        self._host = host
+        self._github_project = github_project
+
+    @guard('search')
+    def get_pull_requests(self, updated_since):
+        window_start = time.strftime('%Y-%m-%dT%H:%M:%SZ', updated_since)
+        url = '{}/search/issues?q=repo:{}+is:pr+updated:>{}'.format(
+            self._host, self._github_project, window_start
+        )
+
+        logger.info(
+            'Searching for Pull Requests updated since %s', window_start
+        )
+
+        data = gh_request('GET', url)
+
+        logger.info('Found %d Pull Requests', len(data['items']))
+
+        if data['incomplete_results']:
+            raise Exception('Incomplete results')
+
+        return data['items']
+
+    @guard('core')
+    def create_ref(self, refspec, revision):
+        url = '{}/repos/{}/git/refs'.format(self._host, self._github_project)
+
+        logger.info('Creating ref "%s" (%s)', refspec, revision)
+
+        gh_request('POST', url, {
+            'ref': 'refs/{}'.format(refspec),
+            'sha': revision
+        })
+
+    @guard('core')
+    def update_ref(self, refspec, revision):
+        url = '{}/repos/{}/git/refs/{}'.format(
+            self._host, self._github_project, refspec
+        )
+
+        logger.info('Updating ref "%s" (%s)', refspec, revision)
+
+        gh_request('PATCH', url, {'sha': revision})
+
+    @guard('core')
+    def create_deployment(self, pull_request, revision):
+        url = '{}/repos/{}/deployments'.format(
+            self._host, self._github_project
+        )
+        # The Pull Request preview system only exposes one Deployment for a
+        # given Pull Request. Identifying the Deployment by the Pull Request
+        # number ensures that GitHub.com automatically responds to new
+        # Deployments by designating prior Deployments as "inactive"
+        environment = 'gh-{}'.format(pull_request['number'])
```

Comment:
Does the 'gh-' prefix here have meaning to the GitHub APIs, or is it to uniquely identify PR preview deployments? If the latter, we may want to use something more specific.

Reply:
Good idea.

Making this change surfaced an integration bug. Since I wrote this script, [we changed the URLs generated by the preview server](https://github.com/bocoup/web-platform-tests.live/pull/23), but I forgot to accommodate that change in the logic that detects preview status. I've corrected that, too.